### PR TITLE
🎉 Source Klaviyo: Add Profiles Stream

### LINKED ISSUE #17393: Source Klaviyo: new stream Profiles
## Tell us about the problem you're trying to solve
[Profiles](https://developers.klaviyo.com/en/reference/get-profile)
Returns a list of all the email templates you've created. The templates are returned in sorted order by _name_.

## Describe the solution you’d like
Research pros and cons for implementing this stream.
_Research should contain: what stream should be implemented, permissions and scopes (if needed), paid plan (if needed)._

### PR DESCRIPTION
**Pull request by Wittiest**
## What
- Add incremental Profiles stream to source Klaviyo
- Closes #17393
- Klaviyo has introduced a new API version. While we generally need to migrate the old endpoints to this API version, this PR is only implementing the Profiles API. Followup PRs can migrate the remaining endpoints.

## How
- Introduce base classes for working with the new Klaviyo API which has cursor-based pagination
  - Note: You can see in `#next_page_token` there is some odd parameter logic. I adopted this from the Orbit connector code here: https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-orbit/source_orbit/streams.py#L36 since it seemed to be a fine solution. I am thinking after there are more cases of APIs using the [JSON API Spec](https://jsonapi.org/), there could be value in some standardized handling of these APIs
- Introduce Profiles API relying on IncrementalKlaviyoStreamLatest
- Rename old APIs to indicate that they are deprecated

## 🚨 User Impact 🚨
- Add incremental Profiles stream to source Klaviyo

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by Wittiest** [comment]
**Unit Tests Passing**
<details>
    <summary>View Screenshot</summary>
Screen Shot 2023-03-13 at 5 37 24 AM
</details>

**Comment by Wittiest** [comment]
**Screenshot of Flake Tests Passing**
<details>
    <summary>View Screenshot</summary>
Screenshot 2023-03-14 at 9 04 20 AM

</details>

@sajarin I believe the latest commit has fixed all issues. Can you rerun integration tests?

**Comment by igrankova** [comment]
@YowanR can you help Dustin to speed up this PR testing and reviewing?

**Comment by Wittiest** [comment]
@sajarin Huge apologies! I failed to update the unit tests. I have tested again locally and all seems to be good now:

<details>
    <summary>View Output</summary>

```
(.venv) ➜  source-klaviyo git:(klaviyo-profiles-stream) ✗ python -m pytest unit_tests
Test session starts (platform: darwin, Python 3.9.14, pytest 6.2.5, pytest-sugar 0.9.6)
cachedir: .pytest_cache
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/Users/dustinpearson/airbyte/airbyte-integrations/connectors/source-klaviyo/.hypothesis/examples')
rootdir: /Users/dustinpearson/airbyte, configfile: pytest.ini
plugins: hypothesis-6.54.6, requests-mock-1.9.3, mock-3.6.1, sugar-0.9.6, timeout-1.4.2, cov-3.0.0
collecting ... 
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_latest_streams.py::TestIncrementalKlaviyoStreamLatest.test_cursor_field_is_required ✓     3% ▍         
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_latest_streams.py::TestIncrementalKlaviyoStreamLatest.test_next_page_token[response_json0-next_page_token0] ✓5% ▌         
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_latest_streams.py::TestProfilesStream.test_parse_response ✓                               8% ▊         
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_source.py::test_check_connection[200--True-None] ✓                                       11% █▏        {"type": "LOG", "log": {"level": "ERROR", "message": "{}"}}

 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_source.py::test_check_connection[400-Bad request-False-HTTPError('400 Client Error: None for url: https://a.klaviyo.com/api/v1/metrics?api_key=api_key&count=100')] ✓13% █▍        {"type": "LOG", "log": {"level": "ERROR", "message": "{}"}}

 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_source.py::test_check_connection[403-Forbidden-False-HTTPError('403 Client Error: None for url: https://a.klaviyo.com/api/v1/metrics?api_key=api_key&count=100')] ✓16% █▋        
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_next_page_token[response_json0-next_page_token0] ✓  18% █▉        
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_next_page_token[response_json1-next_page_token1] ✓  21% ██▏       
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_next_page_token[response_json2-None] ✓              24% ██▍       
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_request_params[next_page_token0-expected_params0] ✓ 26% ██▋       
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_request_params[None-expected_params1] ✓             29% ██▉       
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestKlaviyoStreamV1.test_parse_response ✓                                    32% ███▎      
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_cursor_field_is_required ✓               34% ███▌      
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[None-stream_state0-expected_params0] ✓37% ███▊      
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[next_page_token1-stream_state1-expected_params1] ✓39% ███▉      
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[None-stream_state2-expected_params2] ✓42% ████▎     
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[next_page_token3-stream_state3-expected_params3] ✓45% ████▌     
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[next_page_token4-stream_state4-expected_params4] ✓47% ████▊     
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_request_params[None-stream_state5-expected_params5] ✓50% █████     
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_get_updated_state[current_state0-latest_record0-expected_state0] ✓53% █████▍    
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_get_updated_state[current_state1-latest_record1-expected_state1] ✓55% █████▌    
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_get_updated_state[current_state2-latest_record2-expected_state2] ✓58% █████▊    
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_get_updated_state[current_state3-latest_record3-expected_state3] ✓61% ██████▏   
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_next_page_token[response_json0-next_page_token0] ✓63% ██████▍   
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestIncrementalKlaviyoStreamV1.test_next_page_token[response_json1-None] ✓   66% ██████▋   
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_cursor_field_is_required ✓        68% ██████▉   
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_state_checkpoint_interval ✓       71% ███████▏  
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_request_params[None-stream_state0-expected_params0] ✓74% ███████▍  
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_request_params[next_page_token1-stream_state1-expected_params1] ✓76% ███████▋  
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_request_params[None-stream_state2-expected_params2] ✓79% ███████▉  
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_request_params[next_page_token3-stream_state3-expected_params3] ✓82% ████████▎ 
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_get_updated_state[current_state0-latest_record0-expected_state0] ✓84% ████████▌ 
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_get_updated_state[current_state1-latest_record1-expected_state1] ✓87% ████████▊ 
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_get_updated_state[current_state2-latest_record2-expected_state2] ✓89% ████████▉ 
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_next_page_token ✓                 92% █████████▎
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_parse_response_read_backward ✓    95% █████████▌
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestReverseIncrementalKlaviyoStreamV1.test_parse_response_read_forward ✓     97% █████████▊
 airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py::TestEventsStream.test_parse_response ✓                                      100% ██████████
========================================================================== warnings summary ===========================================================================
airbyte-integrations/connectors/source-klaviyo/unit_tests/test_latest_streams.py: 2 warnings
airbyte-integrations/connectors/source-klaviyo/unit_tests/test_source.py: 3 warnings
airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py: 30 warnings
  /Users/dustinpearson/airbyte/airbyte-integrations/connectors/source-klaviyo/.venv/lib/python3.9/site-packages/airbyte_cdk/sources/streams/http/http.py:45: DeprecationWarning: Call to deprecated class NoAuth. (Set `authenticator=None` instead) -- Deprecated since version 0.1.20.
    self._authenticator: HttpAuthenticator = NoAuth()

airbyte-integrations/connectors/source-klaviyo/unit_tests/test_latest_streams.py: 2 warnings
airbyte-integrations/connectors/source-klaviyo/unit_tests/test_source.py: 3 warnings
airbyte-integrations/connectors/source-klaviyo/unit_tests/test_streams.py: 30 warnings
  /Users/dustinpearson/airbyte/airbyte-integrations/connectors/source-klaviyo/.venv/lib/python3.9/site-packages/deprecated/classic.py:173: DeprecationWarning: Call to deprecated class HttpAuthenticator. (Use requests.auth.AuthBase instead) -- Deprecated since version 0.1.20.
    return old_new1(cls, *args, **kwargs)

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (0.29s):
      38 passed

```

</details>

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/schemas/profiles.json [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        "last_event_date": { "type": "string", "format": "date-time" }
+      }
+    },
+    "properties": { "type": "object" },
+    "subscriptions": { "type": "object" },
+    "links": { "type": "object" },
+    "relationships":  { "type": "object" },
+    "segments":  { "type": "object" }
+  }
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/schemas/profiles.json:26]: EOF and please format the file according to other schemas.
The convention is to use `["null", "type"]` to not have problems with missing data. Check other streams

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py [hunk 2]
```diff
@@ -13,8 +14,133 @@
...
+        """
+        decoded_response = response.json()
+
+        links = decoded_response.get("links")
+        if not links:
+            return None
+
+        next = links.get("next")
+        if not next:
+            return None
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/source_klaviyo/streams.py:58]:
```suggestion
        links = decoded_response.get("links", {})
        next = links.get("next")
        if not next:
            return None
```

File: airbyte-integrations/connectors/source-klaviyo/acceptance-test-config.yml [hunk 3]
```diff
@@ -5,6 +5,8 @@ acceptance_tests:
       empty_streams:
       - name: flows
         bypass_reason: "the 'trigger' object may mutate often"
+      - name: profiles
+        bypass_reason: "the airbyte team will need to add these to integration tests"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-klaviyo/acceptance-test-config.yml:9]: @Wittiest for this change please remove this line. Integration tests must validate the new stream is working.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
@Wittiest / @YowanR integration tests are passing because the new stream was bypassed during tests. Please check some comments and turn the `profiles` stream on in acceptance test config file.

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py [hunk 4]
```diff
@@ -25,10 +27,11 @@ def check_connection(self, logger, config: Mapping[str, Any]) -> Tuple[bool, Any
         return True, None
 
     def streams(self, config: Mapping[str, Any]) -> List[Stream]:
-        """
-        Discovery method, returns available streams
+        """Discovery method, returns available streams
```
[the remark this answers is not in the document]
Comment (Wittiest) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py:30]: Hi @marcosmarxm, can you add more detail? When I ran the linter, it forced me to modify the string to a new format even though I didn't change this part of the code.

Is there another linter I am missing that you can help me to run?

[the remark this answers is not in the document]
Comment (marcosmarxm) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py:30]: What command are you running with linter? When you run `./gradlew ...:integratinoTest` Airbyte linter is applied.

[the remark this answers is not in the document]
Comment (Wittiest) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/source.py]: Hi @marcosmarxm. I was running `./gradlew --no-daemon :airbyte-integrations:connectors:source-klaviyo:airbytePythonFormat`, but it appears I should have run `./gradlew --no-daemon :airbyte-integrations:connectors:source-klaviyo:pipInstall` first. Somehow it was giving different output.

I've reverted this change and many others that were made based on that invalid output

**Comment by Wittiest** [comment]
@marcosmarxm 
> Failed: High strictness level error: profiles streams are declared in the catalog but do not have expected records. Please add expected records to integration_tests/expected_records.jsonl or declare these streams in empty_streams.

Will a member of the airbyte team be able to do this?

**Comment by marcosmarxm** [comment]
1st sync using current master version
2nd sync using the new version with profiles
image

**Review by marcosmarxm — APPROVED** [review]
Thanks @Wittiest this is awesome!

**Review by lazebnyi — APPROVED** [review]

**Review by lazebnyi — APPROVED** [review]
LGTM! Test passed, expected records present.

**Comment by Wittiest** [comment]
@marcosmarxm @lazebnyi , we have a bug in this PR so it probably shouldn't be merged.

We had an ongoing thread [here](https://airbytehq.slack.com/archives/C021JANJ6TY/p1679071082986879) where we were confused about why data wasn't being displayed properly with normalization.

It seems that the **properties** response from the API is somehow being automatically nested underneath **attributes** because of **properties being a keyword**

Example API Response:

  ```json
{
   "attributes": {
      "email": "jpires11@icloud.com",
    },
    "properties": {
      "PROPERTY1": "test",
    }
}
  ```
  
Example of how this is being stored:

```json
{
   "attributes": {
      "email": "jpires11@icloud.com",
      "properties": {
        "PROPERTY1": "test",
      }
    },
}
```

**We will need to think of a solution to this problem.**

A simple solution could be to load the _properties_ data into a new field with a different name at the top-level, but maybe someone has a better idea?

File: airbyte-integrations/connectors/source-klaviyo/source_klaviyo/schemas/profiles.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    "updated": { "type": ["null", "string"], "format": "date-time" },
+    "attributes": {
+      "type": ["null", "object"],
+      "additionalProperties": true,
+      "properties": {
+        "email": { "type": ["null", "string"] },
+        "phone_number": { "type": ["null", "string"] },
+        "first_name": { "type": ["null", "string"] },
+        "last_name": { "type": ["null", "string"] },
+        "organization": { "type": ["null", "string"] },
```
[the remark this answers is not in the document]
Comment (Wittiest) [reply airbyte-integrations/connectors/source-klaviyo/source_klaviyo/schemas/profiles.json:16]: @marcosmarxm You are right. This entire time I wasn't reading the JSON response properly from the API and was thinking this was adjacent. I agree we should move this field

**Review by marcosmarxm — CHANGES_REQUESTED** [review]

**Review by marcosmarxm — APPROVED** [review]

**Pull request merged**